[#7119] Improvement(dev): Support environment variable injection for gravitino.conf

KyleLin0927 opened the pull request:
### What changes were proposed in this pull request?

Enhanced the `dev/docker/gravitino` Docker image to support environment variable injection for `gravitino.conf` configuration. This change aligns the Gravitino server image behavior with the existing design pattern used by `iceberg-rest-server`.

- Added a new `rewrite_config.py` script that reads predefined environment variables and updates `gravitino.conf` at container startup.
- Added `start-gravitino.sh` to execute `rewrite_config.py` before launching the Gravitino server.

### Why are the changes needed?

Fixes #7119

### Does this PR introduce any user-facing change?

This is an internal improvement to the `dev/docker/gravitino` image and does not affect Gravitino core functionality.

### How was this patch tested?

- Ran `./gradlew clean build`.
- Executed `dev/docker/gravitino/gravitino-dependency.sh` to verify the dependency script works properly and built the Gravitino image.
- Launched a basic container and confirmed the Gravitino server started successfully:
    
    ```bash
     ./dev/docker/build-docker.sh --platform all --type gravitino --image kylelin0927/gravitino --tag test-rewrite-config
    ```

Verified environment variable injection for key settings:

- Test `GRAVITINO_SERVER_WEBSERVER_HTTP_PORT`: Server was reachable on port 8080.
    
    ```bash
    docker run --rm -d \
      -e GRAVITINO_SERVER_WEBSERVER_HTTP_PORT=8080 \
      -p 8080:8080 \
      kylelin0927/gravitino:test-rewrite-config
    ```
    
- Test `GRAVITINO_ENTITY_STORE_RELATIONAL_JDBC_URL` and `GRAVITINO_ENTITY_STORE_RELATIONAL_JDBC_DRIVER`: Server started successfully without error logs.
    
    ```bash
    docker run --rm -d \
      -e GRAVITINO_ENTITY_STORE_RELATIONAL_JDBC_URL="jdbc:postgresql://localhost:5432/database1" \
      -e GRAVITINO_ENTITY_STORE_RELATIONAL_JDBC_DRIVER="org.postgresql.Driver" \
      -p 8090:8090 \
      kylelin0927/gravitino:test-rewrite-config
    ```
    
- Tested `GRAVITINO_SERVER_WEBSERVER_MIN_THREADS` and `GRAVITINO_SERVER_WEBSERVER_MAX_THREADS`: verified that `/root/gravitino/conf/gravitino.conf` contains the correct values using cat and grep.
    
    ```bash
    docker run -d --rm --name gravitino-test \
      -e GRAVITINO_SERVER_WEBSERVER_MIN_THREADS=10 \
      -e GRAVITINO_SERVER_WEBSERVER_MAX_THREADS=50 \
      kylelin0927/gravitino:test-rewrite-config
    ```

unknowntpo commented:
@KyleLin0927 In your test plan. Docker image should be built using this build script, it will automatically fetch dependencies and execute docker build.

```
 ./dev/docker/build-docker.sh --platform all --type gravitino --image <image name> --tag <tag name>
```

For example:

```
 ./dev/docker/build-docker.sh --platform all --type gravitino --image unknowntpo/gravitino --tag test-rewrite-config
```

KyleLin0927 commented:
@unknowntpo  Thanks for the suggestion! I tried it out with ./dev/docker/build-docker.sh --platform all --type gravitino --image kylelin0927/gravitino --tag test-rewrite-config, pulled the image, and went ahead with the env variable setup.

FANNG1 commented on dev/docker/gravitino/rewrite_config.py:
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
```
Could you rename this file to `rewrite_gravitino_server_config.py` to distinguish with `rewrite_config.py` in  Gravitino Iceberg server docker image ?

KyleLin0927 replied:
updated it.

unknowntpo commented on dev/docker/gravitino/rewrite_config.py:
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
```
This is not a bash script, and this python file is executed by `python` command, so we can remove this line.

KyleLin0927 replied:
Thanks for the reminder. it's updated now.

unknowntpo commented:
@KyleLin0927 In the test plan, `GRAVITINO_ENTITY_STORE_JDBC_URL` should be `GRAVITINO_ENTITY_STORE_RELATIONAL_JDBC_URL`, which mapped to `gravitino.entity.store.relational.jdbcUrl` (you forgot `RELATIONAL`)

Except this flaw, I've followed your test plan and successfully reproduced all steps with correct results.

KyleLin0927 commented:
I have updated commit:

1. Renamed the file to `rewrite_gravitino_server_config.py` @
2. Removed the unnecessary hashbang
3. Double-checked the env vars — including `GRAVITINO_ENTITY_STORE_RELATIONAL_JDBC_URL` and other DB config variable.

unknowntpo commented:
LGTM, @FANNG1 WDYT?

FANNG1 commented:
> LGTM, @FANNG1 WDYT?

I will do some tests latter.

FANNG1 commented:
Could add related document in `https://gravitino.apache.org/docs/0.9.0-incubating/gravitino-server-config`  like `https://gravitino.apache.org/docs/0.9.0-incubating/gravitino-server-config#how-to-set-up-runtime-environment-variables`  ?

KyleLin0927 commented:
> Could you add related document in `https://gravitino.apache.org/docs/0.9.0-incubating/gravitino-server-config`  like `https://gravitino.apache.org/docs/0.9.0-incubating/gravitino-server-config#how-to-set-up-runtime-environment-variables`  ?

Sure. I will give a try later today.

KyleLin0927 commented:
> Could you add related document in `https://gravitino.apache.org/docs/0.9.0-incubating/gravitino-server-config` like `https://gravitino.apache.org/docs/0.9.0-incubating/gravitino-server-config#how-to-set-up-runtime-environment-variables` ?

I have edited doc. @FANNG1 plz take a look, thanks

KyleLin0927 commented on docs/gravitino-server-config.md:
```diff
@@ -259,7 +259,54 @@ You could put HDFS configuration file to the catalog properties configuration di
+
+Edit the `gravitino-env.sh` script to set environment variables for local development.
+
+### Docker-based Environment Variable Injection
```
I have updated the doc and referred to the Iceberg REST document.

FANNG1 commented:
merged to main, thanks @KyleLin0927  for your contribution and @unknowntpo for your review!
